pkg/storage/storage_fs.go

```diff
@@ -0,0 +1,1340 @@
+	if err != nil {
+		is.log.Error().Err(err).Msg("failure walking storage root-dir")
+		return nil, err
+	}
+
+	stores := make([]string, 0)
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !info.IsDir() {
+			return nil
+		}
+
+		rel, err := filepath.Rel(is.rootDir, path)
+		if err != nil {
+			return nil
+		}
+
+		if ok, err := is.ValidateRepo(rel); !ok || err != nil {
+			return nil
+		}
+
+		//is.log.Debug().Str("dir", path).Str("name", info.Name()).Msg("found image store")
+		stores = append(stores, rel)
+
+		return nil
+	})
+
+	return stores, err
+}
+
+// GetImageTags returns a list of image tags available in the specified repository.
+func (is *ImageStoreFS) GetImageTags(repo string) ([]string, error) {
+	dir := path.Join(is.rootDir, repo)
+	if !dirExists(dir) {
+		return nil, errors.ErrRepoNotFound
+	}
+
+	is.RLock()
+	defer is.RUnlock()
+
+	buf, err := ioutil.ReadFile(path.Join(dir, "index.json"))
+	if err != nil {
+		is.log.Error().Err(err).Str("dir", dir).Msg("failed to read index.json")
+		return nil, errors.ErrRepoNotFound
+	}
+
+	var index ispec.Index
+	if err := json.Unmarshal(buf, &index); err != nil {
+		is.log.Error().Err(err).Str("dir", dir).Msg("invalid JSON")
+		return nil, errors.ErrRepoNotFound
+	}
+
+	tags := make([]string, 0)
+
+	for _, manifest := range index.Manifests {
+		v, ok := manifest.Annotations[ispec.AnnotationRefName]
+		if ok {
+			tags = append(tags, v)
+		}
+	}
+
+	return tags, nil
+}
+
+// GetImageManifest returns the image manifest of an image in the specific repository.
+func (is *ImageStoreFS) GetImageManifest(repo string, reference string) ([]byte, string, string, error) {
+	dir := path.Join(is.rootDir, repo)
+	if !dirExists(dir) {
+		return nil, "", "", errors.ErrRepoNotFound
+	}
+
+	is.RLock()
+	defer is.RUnlock()
+
+	buf, err := ioutil.ReadFile(path.Join(dir, "index.json"))
+
+	if err != nil {
+		is.log.Error().Err(err).Str("dir", dir).Msg("failed to read index.json")
+
+		if os.IsNotExist(err) {
+			return nil, "", "", errors.ErrRepoNotFound
+		}
+
+		return nil, "", "", err
+	}
+
+	var index ispec.Index
+	if err := json.Unmarshal(buf, &index); err != nil {
+		is.log.Error().Err(err).Str("dir", dir).Msg("invalid JSON")
+		return nil, "", "", err
+	}
+
+	found := false
+
+	var digest godigest.Digest
+
+	mediaType := ""
+
+	for _, m := range index.Manifests {
+		if reference == m.Digest.String() {
+			digest = m.Digest
+			mediaType = m.MediaType
+			found = true
+
+			break
+		}
+
+		v, ok := m.Annotations[ispec.AnnotationRefName]
+		if ok && v == reference {
+			digest = m.Digest
+			mediaType = m.MediaType
+			found = true
+
+			break
+		}
+	}
+
+	if !found {
+		return nil, "", "", errors.ErrManifestNotFound
+	}
+
+	p := path.Join(dir, "blobs", digest.Algorithm().String(), digest.Encoded())
+
+	buf, err = ioutil.ReadFile(p)
+
+	if err != nil {
+		is.log.Error().Err(err).Str("blob", p).Msg("failed to read manifest")
```

Comment:
this can be covered by generating a fake image which does not contain manifest blob.